src/rust/engine/process_execution/src/lib.rs

```diff
@@ -275,6 +276,51 @@ impl InputDigests {
     })
   }
 
+  pub async fn new_from_merged(store: &Store, from: Vec<InputDigests>) -> Result<Self, String> {
+    Ok(Self {
+      complete: store
+        .merge(
+          from
+            .iter()
+            .map(|input_digests| input_digests.complete.clone())
+            .collect(),
+        )
+        .await?,
```

Comment:
These sequential `await`s will run sequentially.

To have them run in parallel, you should use `future::join!`: https://docs.rs/futures/latest/futures/macro.join.html